File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/digitaltwin/element/ReferenceElementType.java
package org.opcfoundation.webserver.digitaltwin.element;

import org.eclipse.milo.opcua.stack.core.NodeIds;
import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.eclipse.milo.opcua.stack.core.types.builtin.Variant;
import org.eclipse.milo.opcua.stack.core.types.enumerated.NodeClass;
import org.opcfoundation.webserver.addressspace.nodes.*;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaObjectTypes;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaReferenceTypes;
import org.opcfoundation.webserver.digitaltwin.DigitalTwinSpace;
import org.opcfoundation.webserver.digitaltwin.callback.ReferenceElementCallback;
import org.opcfoundation.webserver.service.message.digitaltwin.GetDescriptorRequest;
import org.opcfoundation.webserver.types.digitaltwin.ReferenceTargetDescriptor;
import org.opcfoundation.webserver.types.digitaltwin.ObjectServiceContext;
import org.opcfoundation.webserver.types.common.UaBrowseAdditionalInfo;
import org.opcfoundation.webserver.types.common.UaReferenceDescriptor;
import org.opcfoundation.webserver.service.message.reactiveobject.BrowseObjectRequest;
import org.opcfoundation.webserver.service.message.reactiveobject.BrowseObjectResponse;
import org.opcfoundation.webserver.service.message.digitaltwin.GetLinkRequest;
import org.opcfoundation.webserver.service.message.digitaltwin.GetLinkResponse;
import org.opcfoundation.webserver.service.message.reactiveobject.ReadObjectAttributeRequest;
import org.opcfoundation.webserver.service.message.reactiveobject.ReadObjectAttributeResponse;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public abstract class ReferenceElementType extends ElementType implements ReferenceElementCallback {
    public ReferenceElementType(
            String typeId,
            LocalizedText displayName,
            DigitalTwinSpace twinSpace)
    {
        super(
                typeId,
                displayName,
                UaObjectTypes.ReferenceElementType,
                twinSpace);
    }

    @Override
    public final boolean isGetLinkSupported()
    {
        return true;
    }

    @Override
    public final CompletableFuture<ReadObjectAttributeResponse> onReadObjectAttributes(ReadObjectAttributeRequest request)
    {
        ObjectServiceContext context = new ObjectServiceContext(request.getObjectId());
        GetDescriptorRequest getDescriptorRequest = new GetDescriptorRequest(context);

        return onGetDescriptor(getDescriptorRequest).thenApply(response -> {
            return new ReadObjectAttributeResponse(request.getObjectId(), response.getDisplayName(), response.getDescription());
        });
    }

    @Override
    public final CompletableFuture<BrowseObjectResponse> onBrowseObjectLinks(BrowseObjectRequest request)
    {
        if (!NodeIds.References.equals(request.getBrowseDescription().getReferenceTypeId()) &&
                !NodeIds.NonHierarchicalReferences.equals(request.getBrowseDescription().getReferenceTypeId()))
        {
            return CompletableFuture.completedFuture(new BrowseObjectResponse(new ArrayList<>(), false));
        }

        ObjectServiceContext context = new ObjectServiceContext(request.getObjectId());
        GetLinkRequest getLinkRequest = new GetLinkRequest(
                context,
                request.getAdditionalInfo().getMaxReferencesPerNode(),
                request.getAdditionalInfo().getReferenceOffset());

        return onGetLinks(getLinkRequest).
                thenApply(this::processBrowseLinkResponse);
    }

    private BrowseObjectResponse processBrowseLinkResponse(GetLinkResponse response) {
        List<UaReferenceDescriptor> linkDescriptors = new ArrayList<>();

        for (ReferenceTargetDescriptor item: response.getTargets())
        {
            UaReferenceDescriptor descriptor;

            if (null == item.getInstanceDeclaration())
            {
                descriptor = new UaReferenceDescriptor(
                        item.getId(),
                        NodeClass.Object,
                        item.getId(),
                        item.getDisplayName(),
                        item.getTypeId(),
                        UaReferenceTypes.HasLink.nodeId(),
                        true);
            } else {
                descriptor = new UaReferenceDescriptor(
                        item.getId(),
                        item.getInstanceDeclaration(),
                        UaReferenceTypes.HasLink.nodeId(),
                        true);
            }

            linkDescriptors.add(descriptor);
        }

        return new BrowseObjectResponse(linkDescriptors, response.containsMoreData(), UaBrowseAdditionalInfo.GET_LINK_TASK);
    }
}

File: opcua-webservice-node/source/src/server/digitaltwin/element/ReferenceElementType.ts
import { NodeClass } from "opcua-webapi";
import { ReferenceTypeIds, UaLocalizedText, UaNodeId } from "opcua-webapi-ts";
import { UaObjectTypes } from "../../addressspace/nodes/builtin";
import { BrowseObjectRequest, BrowseObjectResponse, GetLinkRequest, GetLinkResponse } from "../../service/message";
import { UaBrowseAdditionalInfo, UaReferenceDescriptor } from "../../types";
import { ObjectServiceContext } from "../../types/digitaltwin/ObjectServiceContext";
import { ElementType } from "./ElementType";
import { DigitalTwinSpace } from "../DigitalTwinSpace";

export abstract class ReferenceElementType extends ElementType {
    constructor(
        typeId: string,
        displayName: UaLocalizedText,
        twinSpace: DigitalTwinSpace) {
        super(typeId, displayName, UaObjectTypes.ReferenceElementType, twinSpace);
    }

    /**
     * Override in subclasses to return linked objects.
     */
    abstract onGetLinks(request: GetLinkRequest): Promise<GetLinkResponse>;   

    /**
     * Internal framework callback used by the base type to get the reference type id for this repository.
     * Do not call or override this method directly.
     */
    supportedReferenceType(): UaNodeId {
        return UaNodeId.from(ReferenceTypeIds.Organizes);
    }

    /**
     * Internal framework callback used by the base type to browse linked objects.
     * Do not call or override this method directly.
     */
    override async onBrowseObject(request: BrowseObjectRequest): Promise<BrowseObjectResponse> {
        if (!request.additionalInfo.isTaskRequired(UaBrowseAdditionalInfo.GET_RELATED_OBJECT_TASK)) {
            return new BrowseObjectResponse([]);
        }

        const context = new ObjectServiceContext(request.objectId);
        const response = await this.onGetLinks(
            new GetLinkRequest(
                context,
                request.additionalInfo.maxReferencesPerNode,
                request.additionalInfo.referenceOffset,
            ),
        );

        return this.processBrowseObjectResponse(response,request.additionalInfo);
    }

    private processBrowseObjectResponse(
        response: GetLinkResponse,
        additionalInfo: UaBrowseAdditionalInfo): BrowseObjectResponse {
        const linkDescriptors: Array<UaReferenceDescriptor> = [];
        const referenceType = this.supportedReferenceType();

        for (const item of response.targets) {
            const descriptor = new UaReferenceDescriptor(
                    item.id,
                    NodeClass.Object,
                    item.id,
                    item.displayName,
                    item.typeId,
                    referenceType,
                    item.instanceDeclaration === null ? UaNodeId.nullNodeId : item.instanceDeclaration.nodeId);

            linkDescriptors.push(descriptor);
        }

        let taskMask = (response.containsMoreData) ? UaBrowseAdditionalInfo.GET_RELATED_OBJECT_TASK : 0;
        return new BrowseObjectResponse(linkDescriptors, taskMask, additionalInfo.referenceOffset + response.targets.length);
    }
}